tools/util_linux.go

```diff
@@ -48,15 +50,32 @@ func CheckCloneFileSupported(dir string) (supported bool, err error) {
 }
 
 func CloneFile(writer io.Writer, reader io.Reader) (bool, error) {
-	fdst, fdstFound := writer.(*os.File)
-	fsrc, fsrcFound := reader.(*os.File)
-	if fdstFound && fsrcFound {
-		if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, fdst.Fd(), BtrfsIocClone, fsrc.Fd()); err != 0 {
+	dstFD, dstIsFile := writer.(*os.File)
+	srcFD, srcIsFile := reader.(*os.File)
+	if !(dstIsFile && srcIsFile) {
+		return false, nil
+	}
+
+	// Try FICLONE, this supported since kernel 4.5.
+	// This can work with XFS, Btrfs and other format which support copy on write.
+	err := syscall.Syscall(syscall.SYS_IOCTL, dstFD.Fd(), C.FICLONE, srcFD.Fd())
+	if err == nil {
+		return true, nil // success
+	} else { // error case
+		if err != C.EINVAL {
```

Comment:
Can we move these into one else-if statement?  That would keep things tidy.

We could even bring the `BTRFS_IOC_CLONE` case up as an additional else-if case, although that might require that we switch from `err == 0` back to `err != 0`.

Reply:
Sorry I miss-understand about `FICLONE` and `BTRFS_IOC_CLONE`.
I updated my PR, could you review again. Thank you.